src/Identity/IdentityServer/DeviceValidator.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System.ComponentModel.DataAnnotations;
+using System.Reflection;
+using Bit.Core.Context;
+using Bit.Core.Entities;
+using Bit.Core.Enums;
+using Bit.Core.Repositories;
+using Bit.Core.Services;
+using Bit.Core.Settings;
+using Duende.IdentityServer.Validation;
+
+namespace Bit.Identity.IdentityServer;
+
+public interface IDeviceValidator
+{
+    Task<Device> SaveDeviceAsync(User user, ValidatedTokenRequest request);
+    Task<bool> KnownDeviceAsync(User user, ValidatedTokenRequest request);
+}
+
+public class DeviceValidator(
+    IDeviceService deviceService,
+    IDeviceRepository deviceRepository,
+    GlobalSettings globalSettings,
+    IMailService mailService,
+    ICurrentContext currentContext) : IDeviceValidator
+{
+    private readonly IDeviceService _deviceService = deviceService;
+    private readonly IDeviceRepository _deviceRepository = deviceRepository;
+    private readonly GlobalSettings _globalSettings = globalSettings;
+    private readonly IMailService _mailService = mailService;
+    private readonly ICurrentContext _currentContext = currentContext;
+
+    /// <summary>
+    /// Save a device to the database. If the device is already known, it will be returned.
+    /// </summary>
+    /// <param name="user">The user is assumed NOT null, still going to check though</param>
+    /// <param name="request">Duende Validated Request that contains the data to create the device object</param>
+    /// <returns>Returns null if user or device is malformed; The existing device if already in DB; a new device login</returns>
+    public async Task<Device> SaveDeviceAsync(User user, ValidatedTokenRequest request)
+    {
+        var device = GetDeviceFromRequest(request);
+        if (device != null && user != null)
+        {
+            var existingDevice = await GetKnownDeviceAsync(user, device);
+            if (existingDevice == null)
+            {
+                device.UserId = user.Id;
+                await _deviceService.SaveAsync(device);
+
+                // This check just seems like a way to avoid sending an email on the first login
+                // maybe vestigial?
```

Comment:
❓ : did we ever discover what this check was doing?

Reply:
No, but based on the code I'm convinced it's to not send a "There was a login on a new device" email to users that have _just_ created an account since it's their first one.

I think this b/c it checks to see if the user is more than 10 minutes old.

Reply:
Ok, do you mind cleaning up that comment to describe that behavior?